SpiceSharp/Simulations/IntegrationMethods/Spice/SpiceIntegrationMethod.cs

```diff
@@ -0,0 +1,341 @@
+﻿using System;
+using System.Collections.Generic;
+using SpiceSharp.Algebra;
+using SpiceSharp.Simulations;
+using SpiceSharp.Simulations.IntegrationMethods.Spice;
+
+namespace SpiceSharp.IntegrationMethods
+{
+    /// <summary>
+    /// The default integration method as implemented by Spice 3f5
+    /// </summary>
+    public abstract class SpiceIntegrationMethod : IntegrationMethod, IBreakpoints
+    {
+        /// <summary>
+        /// The breakpoints
+        /// </summary>
+        public Breakpoints Breakpoints { get; } = new Breakpoints();
+
+        /// <summary>
+        /// True if we just hit a breakpoint earlier
+        /// </summary>
+        public bool Break { get; protected set; }
+
+        /// <summary>
+        /// Transient tolerance correction factor
+        /// </summary>
+        protected double TrTol { get; private set; } = 7.0;
+
+        /// <summary>
+        /// Allowed relative tolerance
+        /// </summary>
+        protected double RelTol { get; private set; } = 1e-3;
+
+        /// <summary>
+        /// Allowed absolute tolerance
+        /// </summary>
+        protected double AbsTol { get; private set; } = 1e-6;
+
+        /// <summary>
+        /// Allowed maximum timestep
+        /// </summary>
+        protected double MaxStep { get; private set; } = 1e-6;
+
+        /// <summary>
+        /// Expansion factor
+        /// </summary>
+        protected double Expansion { get; private set; } = 2.0;
+
+        /// <summary>
+        /// Minimum timestep
+        /// </summary>
+        protected double MinStep { get; private set; }
+
+        /// <summary>
+        /// Gets the prediction vector
+        /// </summary>
+        protected Vector<double> Prediction { get; private set; }
+
+        /// <summary>
+        /// Keep track of all states that can be truncated
+        /// </summary>
+        protected List<ITruncatable> TruncatableStates { get; } = new List<ITruncatable>();
+
+        /// <summary>
+        /// Private variables
+        /// </summary>
+        private double _saveDelta;
+        private double _oldDelta;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        protected SpiceIntegrationMethod(int maxOrder)
+            : base(maxOrder)
+        {
+            TruncateEvaluate += TruncateStates;
+        }
+
+        /// <summary>
+        /// Setup the integration method
+        /// </summary>
+        /// <param name="simulation">Time simulation</param>
+        public override void Setup(TimeSimulation simulation)
+        {
+            base.Setup(simulation);
+
+            // Basic time configuration
+            var tc = simulation.ParameterSets.Get<TimeConfiguration>();
+            Breakpoints.SetBreakpoint(tc.InitTime);
+            Breakpoints.SetBreakpoint(tc.FinalTime);
+            MaxStep = tc.MaxStep;
+            MinStep = tc.DeltaMin;
+            _saveDelta = tc.FinalTime / 50.0;
```

Comment:
@svenboulanger Why 50.0 ? Please provide comment or a const with a good name.

Reply:
There is no real reason for this number. It is just how Spice 3f5 implemented it.
That is why I implemented `SpiceIntegrationMethod`. It separates all things that are just specific to Spice but are not necessary for general integration methods.

In my opinion, it doesn't even influence any part of the simulation. Changing it does not affect the tests in any way. I am leaving it in for possibly obscure compatibility reasons.

Reply:
It only appears to be relevant at t=0 (the first breakpoint). The timestep is calculated as follows:
1. Calculate initial timestep as `initDelta = Min(FinalTime / 50, Step) / 10`
2. Calculate `mt = Min(**_saveDelta**, TimeToNextBreakpoint)`
3. `delta = Min(initDelta, mt / 10)`
4. For the first step `delta /= 10`

That is the only situation where the initial value of _saveDelta is used. It is always overwritten later. Assuming there are no breakpoints, _saveDelta only influences simulation behavior if `_saveDelta < initDelta`, or `FinalTime / 50 < Min(FinalTime / 50, Step)` which is never the case.